feat: Refactor to requests (#32)
decentraland/auth-server: README.md

For example, if the desktop client need to send a transaction, it would create a transaction for the eth_sendTransaction method, and await for the result, which would be a transaction hash, to be returned after the flow is complete.

Some charactistics of requests are:

1. Only one request can exist at a time per connected socket. A new request will invalidate a previous one if it existed.
2. Requests have an expiration, and cannot be consumed after it.
3. If the socket disconnects, any request made by that socket will be deleted.

## Usage

This section will explain the ways in which the service can be used.

Use the web socket library of your choice to connect to this server (https://auth-api.decentraland.org). This one currently uses [Socket.IO](https://socket.io/) which can also be used in a JS client to connect.

The next example will show how a `personal_sign` can be requested by the desktop client.

1. The desktop client has to connect to the auth server through web sockets.

```ts
const socket = io('https://auth-api.decentraland.org')
```

> 2. The desktop client has to send a request message with the method information to the auth server, at the same time, start listening for the response.

Reviewer: Optional suggestion:
```suggestion
2. The desktop client has to send a request message with the method information to the auth server, and wait for the response.
```

```ts
const { requestId, expiration, code } = await socket.emitWithAck('request', {
  method: 'personal_sign',
  params: ['message to sign', 'signer address']
})
```

> The expiration can be used to know when the request will become unnavailable if not consumed before a certain time.

Reviewer: Optional suggestion:
```suggestion
The expiration shows when the request will become unavailable. The request must be consumed before it expires.
```

      if (msg.error) {
        reject(msg.error)
      } else {
        resolve(msg)
      }
    }
  })
})
```

5. Get the `result` and the `sender` from the outcome message and do with them whatever is necessary.

### Authentication Flow

For the sign in flow in the desktop client. we will need to request a special method called `dcl_personal_sign`.

This methods works similarly to `personal_sign` but with a little difference.

For this example we'll be using `ethers v6` and `@dcl/crypto`

1. The desktop client will need to generate and store an epheremeral wallet.

```ts
const ephemeralAccount = ethers.Wallet.createRandom()
```

2. The desktop client has to set a date in which the identity that will be created, expires.

```ts
const expiration = new Date(Date.now() + 24 * 60 * 60 * 1000) // 1 day in the future as an example.
```

3. Generate the ephemeral message to be signed using the address of the ephemeral account and the expiration.

```ts
const ephemeralMessage = Authenticator.getEphemeralMessage(ephemeralAccount.address, expiration)
```

4. Follow the steps decribed on the [Usage](#usage) section, initializing the flow with the following message.

```ts

> socket.emitWithAck('request', {
>   method: 'dcl_personal_sign',
>   params: [ephemeralMessage, code]

Reviewer: Optional suggestion:
```suggestion
await socket.emitWithAck('request', {
  method: 'dcl_personal_sign',
  params: [ephemeralMessage, code]
```

})
```

> As you can see, there is a simple difference with the previous example. That is that personal_sign requires a second parameter that is the address that will sign the message. But we don't know it yet, so only the ephemeral message is sent. The auth dapp will fill the signing address for us.

Reviewer: Optional suggestion:
```suggestion
As you can see, there is a simple difference with the previous example. That is that personal_sign requires a second parameter that is the address that will sign the message, but we don't know it yet, so only the ephemeral message is sent. The auth dApp will fill the signing address for us.
```
